config/addresses.go

```diff
@@ -2,10 +2,12 @@ package config
 
 // Addresses stores the (string) multiaddr addresses for the node.
 type Addresses struct {
-	Swarm          []string // addresses for the swarm to listen on
-	Announce       []string // swarm addresses to announce to the network, if len > 0 replaces auto detected addresses
-	AppendAnnounce []string // similar to Announce but doesn't overwrite auto detected addresses, they are just appended
-	NoAnnounce     []string // swarm addresses not to announce to the network
-	API            Strings  // address for the local API (RPC)
-	Gateway        Strings  // address to listen on for IPFS HTTP object gateway
+	Swarm               []string // addresses for the swarm to listen on
+	Announce            []string // swarm addresses to announce to the network, if len > 0 replaces auto detected addresses
+	AppendAnnounce      []string // similar to Announce but doesn't overwrite auto detected addresses, they are just appended
+	NoAnnounce          []string // swarm addresses not to announce to the network
+	ExtraReframeRouters []string // allows to add other resolvers using the Reframe spec: https://github.com/ipfs/specs/blob/master/REFRAME.md
```

Comment:
The Addresses section is about the nodes addresses not related to the routing system.

An early draft of what this might look like instead is:

```
{
	ReframeRouters : {
		"router-name" : {
			"Address" : "<URL>",
			"Enabled" : boolean,
			"Methods" : ["put-ipns", "get-p2p-providers", ...],
		},
		"additional-router" : {
			"Address" : "<URL>",
			"Enabled" : boolean,
			"Methods" : ["put-ipns", "get-ipns", ...],
		}
	}
}
```

If we were willing to rock the boat more we could also change around the `Routing` type to be more descriptive and encompassing of additional routing systems. This may/may not require a config file migration so has added complexity. Part of the reason to take this on is that if we add support for a new routing system in go-ipfs by default (e.g. the indexer network) that we'll have to add new config options for that anyway. Doing this makes embedding new systems into go-ipfs easier and more explicit.

```
	Routers : {
		"router-name" : {
			"Type" : "reframe"
			"Enabled" : "boolean"
		  "Methods" : ["put-ipns", "get-p2p-providers", ...]
			"Parameters" : {
				"Address" : "<URL>"
				}
		}
		"ipfs-public-dht" : {  // All fields would have defaults most users could ignore
				"Type" : "dht"
				"Enabled" : "true"
				"Methods" : ["put-ipns", "get-p2p-providers", ...]
				"Parameters" : {
					"Mode" : "auto"
					"ProtocolID" : "/ipfs/kad/1.0.0"
					"Bootstrappers" : ....
					"Public-IP-Network" : "true"
				}
		}
		"ipfs-lan-dht" : {
				"Type" : "dht"
				"Enabled" : "false"
				"Methods" : ["put-ipns", "get-p2p-providers", ...]
				"Parameters" : {
					"Mode" : "server"
					"ProtocolID" : "/ipfs/lan/kad/1.0.0"
					"Bootstrappers" : []
					"Public-IP-Network" : "false"
				}
		}
```

cc @lidel @guseggert if you've got any thoughts on how this ends up looking

Reply:
Could you elaborate more about `Methods` param? 
If what we want here is to define the methods that a router will use, maybe would be better to do that at the connection level, returning an error if some method is not implemented on the other side, or implementing some kind of `capabilities` endpoint, where a client can ask for supported methods to the server.

Reply:
Agree that `ExtraReframeRouters` should be moved out of `Addresses`.
Maybe we could move it to already existing `Routing` config section and update `docs/config.md` to reflect that `Routing` supports more than DHT?


We probably want to avoid adding scope to this PR, so (easy) may be good for now, and (hard) can be a follow-up work in separate PR:

* **easy:** 
  - rename `ExtraReframeRouters` to  `Routing.Routers`  – we will use it for future routing systems 
  - change`Routing.Type` to ```*OptionalString  `json:",omitempty"``` – this will stop hardcoding default in config, making it easier for new users to switch to `Routing.Routers` in the future
* **hard:** 
  - same as _easy_ + rock the boat more, make routing config uniform
   - deprecate `Routing.Type` 
      - if user has `Routing.Type` set in their config, print warning asking user to remove   `Routing.Type`  from config and use `Routing.Routers` instead
   - `Routing.Routers` shoud be empty by default, allowing user to override implicit defaults
     - `ipfs-public-dht`, `ipfs-lan-dht` and in the future some "public goods" indexers would be part of implicit defaults 
      - user will be free to override them: values from user config will be applied on top of any implicit default
         - prior  art:  we have implicit default resolver for `.eth`, but user  can override the endpoint (or disable it by setting it to `null`) by defining explicit entry in `DNS.Resolvers`